### PR TITLE
[WIP][mdns] add support for unprovisioned devices in discovery manager

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

We do not support finding unprovisioned devices in DiscoveryManager, which is required for IP rendezvous.

 #### Summary of Changes

Add support for unprovisioned devices.

### CODE REVIEW

File: src/lib/mdns/DiscoveryManager.cpp [hunk 1]
```diff
@@ -330,5 +331,81 @@ void DiscoveryManager::HandleNodeIdResolve(void * context, MdnsService * result,
...
+    else if (result == nullptr)
+    {
+        ChipLogError(Discovery, "Node ID resolve not found");
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceResolve(CHIP_ERROR_UNKNOWN_RESOURCE_ID, 0, MdnsService{});
+    }
+    else
+    {
+        uint64_t name = 0;
+
+        for (size_t i = 0; i < sizeof(uint64_t) * 2; i++)
```
Comment (rwalker-apple) [inline src/lib/mdns/DiscoveryManager.cpp:384]: I think I've seen similar code elsewhere, can we write a utility for this?  I think it should also verify that it's not indexing off the end of result...

**Review by rwalker-apple — CHANGES_REQUESTED** [review]

File: scripts/tests/cirque_tests.sh [hunk 2]
```diff
@@ -56,7 +56,7 @@ function __cirquetest_start_flask() {
     echo 'Start Flask'
     cd "$REPO_DIR"/third_party/cirque/repo
     FLASK_APP='cirque/restservice/service.py' \
-        PATH="$PATH":"$REPO_DIR"/third_party/cirque/repo/openthread/output/x86_64-unknown-linux-gnu/bin/ \
+        PATH="$PATH":"$REPO_DIR"/third_party/cirque/repo/openthread/output/simulation/bin/ \
```
Comment (andy31415) [inline scripts/tests/cirque_tests.sh:59]: Change seems unrelated. Could you merge with master (likely this is already in master)

File: src/lib/mdns/DiscoveryManager.cpp [hunk 3]
```diff
@@ -291,44 +313,100 @@ void DiscoveryManager::HandleNodeIdResolve(void * context, MdnsService * result,
...
         // Parse '%x-%x' from the name
-        uint64_t nodeId       = 0;
-        bool deliminatorFound = false;
+        uint64_t nodeId    = 0;
+        char * deliminator = nullptr;
+
+        HexStringToUint64(result->mName, sizeof(result->mName), &deliminator);
 
-        for (size_t i = 0; i < sizeof(result->mName) && result->mName[i] != 0; i++)
+        if (deliminator == nullptr || *deliminator != '-')
```
Comment (andy31415) [inline src/lib/mdns/DiscoveryManager.cpp:321]: can deliminator be nullptr?

also probably `delimitator` ?

File: src/lib/mdns/DiscoveryManager.cpp [hunk 4]
```diff
@@ -291,44 +313,100 @@ void DiscoveryManager::HandleNodeIdResolve(void * context, MdnsService * result,
...
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceResolve(CHIP_ERROR_UNKNOWN_RESOURCE_ID, 0, MdnsService{});
+    }
+    else
+    {
+        uint64_t name = 0;
+        char * end    = nullptr;
+
+        HexStringToUint64(result->mName, sizeof(result->mName), &end);
+
+        if (end == nullptr || *end != '\0')
```
Comment (andy31415) [inline src/lib/mdns/DiscoveryManager.cpp:388]: same here: I do not think end can be nullptr.

File: src/lib/mdns/DiscoveryManager.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/lib/mdns/DiscoveryManager.cpp:321]: `delimiter`, please....

File: src/lib/mdns/DiscoveryManager.cpp [hunk 5]
```diff
@@ -49,6 +49,27 @@ uint8_t HexToInt(char c)
     return UINT8_MAX;
 }
 
+uint64_t HexStringToUint64(const char * buf, size_t size, char ** end)
+{
+    uint64_t value = 0;
+
+    for (size_t i = 0; i < size; i++)
```
Comment (rwalker-apple) [inline src/lib/mdns/DiscoveryManager.cpp:56]: I guess we don't care about overflow

File: src/lib/mdns/DiscoveryManager.cpp (hunk 3 above)
Comment (rwalker-apple) [reply src/lib/mdns/DiscoveryManager.cpp:321]: delimiter?

File: src/lib/mdns/DiscoveryManager.cpp (hunk 4 above)
Comment (rwalker-apple) [reply src/lib/mdns/DiscoveryManager.cpp:388]: this code doesn't cover another failure mode of `HexStringToUint64()`: when `end == result->mName`, which can occur when the empty string is in `result->mName`.  this code assumes that `result->mName` is always null-terminated.  is that always true?

**Review by rwalker-apple — CHANGES_REQUESTED** [review]

File: src/lib/mdns/DiscoveryManager.cpp [hunk 6]
```diff
@@ -49,6 +49,27 @@ uint8_t HexToInt(char c)
...
+        // This wil also handle buf[i] == '\0'
+        if (current == UINT8_MAX)
+        {
+            *end = const_cast<char *>(&buf[i]);
+            return value;
+        }
+        value = value * 16 + current;
+    }
+
+    *end = const_cast<char *>(&buf[size]);
```
Comment (bzbarsky-apple) [inline src/lib/mdns/DiscoveryManager.cpp:69]: Why do we want this `const_cast`?  Isn't the right solution to make `end` be `const char**`?
